### DISCUSSION TITLE
data modifications

### DISCUSSION THREAD

**Discussion by mauricesar** [Project-MONAI/tutorials#395] [Q&A]
Hi, i'm trying to use Spleen 3D segmentation with MONAI as base for vessel segmentation (task08_hepaticvessel - MSD), do you have any tips for that?

**Comment by Nic-Ma** [comment]
Hi @mauricesar ,

Thanks for your interest here.
I think you can refer to this DynUNet pipeline, it can support all the Decathlon tasks:
https://github.com/Project-MONAI/tutorials/tree/master/modules/dynunet_pipeline

Thanks.

**Reply by mauricesar** [reply]
> Hi @mauricesar ,
> 
> Thanks for your interest here. I think you can refer to this DynUNet pipeline, it can support all the Decathlon tasks: https://github.com/Project-MONAI/tutorials/tree/master/modules/dynunet_pipeline
> 
> Thanks.

Thank you so much for your answer! Is there a possibility for me to use that notebook that i mentioned (Spleen 3D segmentation with MONAI, i'm using google colab) to do this task (vessel segmentation)?

**Reply by Nic-Ma** [reply]
I think you can also modify this tutorial for your task, just need to change below items:
1. input data channels, output prediction channels.
2. Random crop size.

Thanks.

**Comment by mauricesar** [comment]
Hi, i'm having problems with the modification, i keep getting an error: RuntimeError: cuDNN error: CUDNN_STATUS_NOT_INITIALIZED, i tried using another version of torch (multiple people suggested that) but it didn't work. Would you mind taking a look at the code? I'm pretty sure its something really really reeeally simple but I can't fix it. I would be incredible grateful: https://colab.research.google.com/drive/1Ny-ZUa1FOmZDryOAxwrhlxSOlKBQjJ7t?usp=sharing
![Capturar](https://user-images.githubusercontent.com/34106015/137187413-45d4e20f-0cd6-4046-9018-eefb5a7eb531.PNG)

**Reply by ericspod** [reply]
Your network needs to be set to output 3 channels (background, label 1, label 2). It's currently set to 2, this is messing up the dice calculation in a way that Pytorch and MONAI don't check correctly.

**Reply by mauricesar** [reply]
> Your network needs to be set to output 3 channels (background, label 1, label 2). It's currently set to 2, this is messing up the dice calculation in a way that Pytorch and MONAI don't check correctly.

Thank you so so much for your response! You helped me big time! Can I ask you another question? I saw that you commented some parts at the code, were they wrong? I'm getting an error that says: ValueError: The size of the proposed random crop ROI is larger than the image size. Does it have something to do with the spatial_size? (Sorry, English isn't my first language)

**Reply by Nic-Ma** [reply]
Hi @mauricesar ,

About "I saw that you commented some parts at the code, were they wrong?", may I know where the exact code is you are referring to?

Thanks.

**Reply by mauricesar** [reply]
> Hi @mauricesar ,
> 
> About "I saw that you commented some parts at the code, were they wrong?", may I know where the exact code is you are referring to?
> 
> Thanks.

Cell 11, it starts with:

max_epochs = 3
val_interval = 2
best_metric = -1
best_metric_epoch = -1
epoch_loss_values = []
metric_values = []

**Reply by ericspod** [reply]
Oops anything I changed in your notebook should be undone. What I was trying to do was eliminate some variables from the problem one at a time to see what would work, I had thought initially something in the transform chain was wrong. I may have put a break into your training loop as well that should be removed. 

I think the error comes up because your ROI size (96, 96, 96) is maybe larger than some of the input images in at least one dimension.

**Reply by mauricesar** [reply]
> Oops anything I changed in your notebook should be undone. What I was trying to do was eliminate some variables from the problem one at a time to see what would work, I had thought initially something in the transform chain was wrong. I may have put a break into your training loop as well that should be removed.
> 
> I think the error comes up because your ROI size (96, 96, 96) is maybe larger than some of the input images in at least one dimension.

The smallest of the images is (512,512,37). We get: RuntimeError: Sizes of tensors must match except in dimension 4. Got 6 and 5 (The offending index is 0). Any ideas?

**Reply by ericspod** [reply]
Your ROI size is 96 cubed so the minimum dimension of 37 obviously won't work, I think it's that index and no index 0. If your other images are at least 96 in every dimension I would just exclude this one.

**Reply by rijobro** [reply]
You could do `SpatialPadd` to ensure that everything is **at least** `(96, 96, 96)`. Following that you should be able to use `RandCropByPosNegLabeld` with crop size of `(96, 96, 96)`. 

Perhaps it's just whilst you're debugging, but I notice that you're currently using `RandCropByPosNegLabeld` with size 37. However, your network uses `strides=(2, 2, 2, 2)` so your images need to be divisible by 2^4 = 16. So 96 is fine, 37 isn't.

**Reply by mauricesar** [reply]
> You could do `SpatialPadd` to ensure that everything is **at least** `(96, 96, 96)`. Following that you should be able to use `RandCropByPosNegLabeld` with crop size of `(96, 96, 96)`.
> 
> Perhaps it's just whilst you're debugging, but I notice that you're currently using `RandCropByPosNegLabeld` with size 37. However, your network uses `strides=(2, 2, 2, 2)` so your images need to be divisible by 2^4 = 16. So 96 is fine, 37 isn't.

Thanks for your reply! I resized all my images to (160,160,30), when I try to execute the training with max_epochs with a different value (2, 3, 40. The current is 1, i just need the code to work) I get an error saying - RuntimeError: Sizes of tensors must match except in dimension 4. Got 16 and 15 (The offending index is 0).


Do you guys have any ideias about this? What is confusing to me is that I was used to not be able to run this part of the code, and now I can do it but just with 1 epoch. Do you guys have any sugestions about spartial_size and the roi that I should use?

**Reply by rijobro** [reply]
It's hard to say without the full stack trace as I don't know if the problem is in the collation (dataloader) or going through the network. Either way, you can print the dimensions of your data with the transforms applied and ensure they're constant.

(You're also using `set_determinism`, so it's possible that the randomness is such that by chance it works for the first epoch and then fails for the second. Perhaps if you were to remove it and run it multiple times, this wouldn't always be the case.)

**Reply by mauricesar** [reply]
> It's hard to say without the full stack trace as I don't know if the problem is in the collation (dataloader) or going through the network. Either way, you can print the dimensions of your data with the transforms applied and ensure they're constant.
> 
> (You're also using `set_determinism`, so it's possible that the randomness is such that by chance it works for the first epoch and then fails for the second. Perhaps if you were to remove it and run it multiple times, this wouldn't always be the case.)

I tried removing set_determinism, it's still facing problems, here's the full message: 
(the link, if you want to see it: https://colab.research.google.com/drive/1Ny-ZUa1FOmZDryOAxwrhlxSOlKBQjJ7t?usp=sharing)

![Capturar 2](https://user-images.githubusercontent.com/34106015/137775749-65c876da-9fbd-44ad-8996-a333c7c4b148.PNG)

**Reply by rijobro** [reply]
The problem is that your `val_transforms` will produce images of different sizes, since the last crop/pad transformation is `CropForegroundd`, and the background will change for each image. There is then no guarantee that the image will be divisible by factor 16. 

You could add a `DivisiblePadd` after the `CropForegroundd`, or for convenience, `CropForegroundd` has the argument `k_divisible`, which could also set to 16. Let me know if this doesn't work.

**Reply by mauricesar** [reply]
> The problem is that your `val_transforms` will produce images of different sizes, since the last crop/pad transformation is `CropForegroundd`, and the background will change for each image. There is then no guarantee that the image will be divisible by factor 16.
> 
> You could add a `DivisiblePadd` after the `CropForegroundd`, or for convenience, `CropForegroundd` has the argument `k_divisible`, which could also set to 16. Let me know if this doesn't work.

Hi, I'm sorry I took so long to respond - yes, it worked! Thank you so much for your help!

Let me ask you this final question = do you think it is possible for me to give and image as input and then have the segmentation? I want to try other vessels (lungs vessel segmentation, for example) using images that I have (but I don't have their labels). Do you have any tips for this task?
